### ISSUE TITLE
rule proposal: no return in finally

### ISSUE THREAD

**Issue by kentor** [xjamundx/eslint-plugin-promise#44]
`return` inside `finally` is a bit misleading because the return value is _not_ the new resolved value:

```js
// this promise has a resolved value of 1
Promise.resolve(1).finally(() => {
  return 2;
});
```

see: https://github.com/tc39/proposal-promise-finally

**Comment by xjamundx** [comment]
love it

**Comment by thisgeek** [comment]
I've started working on this.

**Comment by xjamundx** [comment]
Thanks!

**Comment by arabold** [comment]
I'm confused on this rule. Shouldn't something like the below be considered valid?

```
Promise.resolve(1).finally(() => {
  return anotherPromise();
});
```

Without the `return` the program wouldn't wait for `anotherPromise()` to finish. Am I wrong?

**Comment by thisgeek** [comment]
@arabold Are you talking about this use case?

> if callback returns a promise, the resolution of the returned promise will be delayed until the promise returned from callback is finished.

— <http://devdocs.io/q/index#promise-finally>

**Comment by arabold** [comment]
Yes, exactly. Sometimes we have cleanup code in the `finally()` section in form of functions that return another promise again. I agree that it makes not much sense to return constants in the `finally()` call, but it should be possible to return a promise in order to have the promise chain wait for it. Does that make sense?

**Comment by thisgeek** [comment]
Would it be worth only registering the error if the argument for the return statement fits certain types? Off the top of my head I'm not sure yet what exactly should be in or out, but for example a `BinaryExpression` should be reported, but a `CallExpression` should not. I guess a whitelist of all the expressions that _could_ result in a promise is a start.

**Comment by gndelia** [comment]
I agree on what's stated here, returning a promise should be considered valid
